checkpoint/side_milestone_handler.go

```diff
@@ -0,0 +1,166 @@
+			"milestoneId", msg.MilestoneID,
+			"error", err,
+		)
+	} else if validMilestone {
+		// vote `yes` if milestone is valid
+		result.Result = abci.SideTxResultType_Yes
+		return
+	}
+
+	logger.Error(
+		"RootHash is not valid",
+		"startBlock", msg.StartBlock,
+		"endBlock", msg.EndBlock,
+		"rootHash", msg.RootHash,
+		"milestoneId", msg.MilestoneID,
+	)
+
+	return common.ErrorSideTx(k.Codespace(), common.CodeInvalidBlockInput)
+}
+
+// PostHandleMsgMilestone handles msg milestone
+func PostHandleMsgMilestone(ctx sdk.Context, k Keeper, msg types.MsgMilestone, sideTxResult abci.SideTxResultType) sdk.Result {
+	logger := k.Logger(ctx)
+	timeStamp := uint64(ctx.BlockTime().Unix())
+
+	// TX bytes
+	txBytes := ctx.TxBytes()
+	hash := tmTypes.Tx(txBytes).Hash()
+
+	// Emit event for milestone
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeMilestone,
+			sdk.NewAttribute(sdk.AttributeKeyAction, msg.Type()),                                  // action
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),                // module name
+			sdk.NewAttribute(hmTypes.AttributeKeyTxHash, hmTypes.BytesToHeimdallHash(hash).Hex()), // tx hash
+			sdk.NewAttribute(hmTypes.AttributeKeySideTxResult, sideTxResult.String()),             // result
+			sdk.NewAttribute(types.AttributeKeyProposer, msg.Proposer.String()),
+			sdk.NewAttribute(types.AttributeKeyStartBlock, strconv.FormatUint(msg.StartBlock, 10)),
+			sdk.NewAttribute(types.AttributeKeyEndBlock, strconv.FormatUint(msg.EndBlock, 10)),
+			sdk.NewAttribute(types.AttributeKeyRootHash, msg.RootHash.String()),
+			sdk.NewAttribute(types.AttributeKeyMilestoneID, msg.MilestoneID),
+		),
+	})
+
+	// Skip handler if milestone is not approved
+	if sideTxResult != abci.SideTxResultType_Yes {
+		logger.Debug("Skipping new milestone since side-tx didn't get yes votes", "startBlock", msg.StartBlock, "endBlock", msg.EndBlock, "rootHash", msg.RootHash, "milestoneId", msg.MilestoneID)
+		k.SetNoAckMilestone(ctx, msg.MilestoneID)
+
+		return sdk.Result{
+			Events: ctx.EventManager().Events(),
+		}
+	}
+
+	if lastMilestone, err := k.GetLastMilestone(ctx); err == nil { // fetch last milestone from store
+		// make sure new milestoen is after tip
+		if lastMilestone.EndBlock > msg.StartBlock {
+			logger.Error(" already exists",
+				"currentTip", lastMilestone.EndBlock,
+				"startBlock", msg.StartBlock,
+			)
+
+			k.SetNoAckMilestone(ctx, msg.MilestoneID)
+
+			return sdk.Result{
+				Events: ctx.EventManager().Events(),
+			}
+		}
+
+		// check if new milestone's start block start from current tip
+		if lastMilestone.EndBlock+1 != msg.StartBlock {
+			logger.Error("milestone not in countinuity",
+				"currentTip", lastMilestone.EndBlock,
+				"startBlock", msg.StartBlock)
+
+			k.SetNoAckMilestone(ctx, msg.MilestoneID)
+
+			return sdk.Result{
+				Events: ctx.EventManager().Events(),
+			}
+		}
+	} else if err != nil && msg.StartBlock != 0 {
```

Comment:
Same with err!=nil. Could you check all places with a few conditions and err!=nil that we check all cases?

Reply:
Err will only occur when db doesn't have milestone